_posts/2021-03-31-relaxing-requirements-of-moved-from-objects.md

```diff
@@ -81,35 +83,45 @@ And from the requirements for the [`assignable_from` concept](https://eel.is/c++
+The last implementation requirement comes from `std::swap()` when invoked as `swap(a, a)`. <!-- I don't see this in the draft; swap is not defined in terms of move AFAICT.  It just says “Effects: Exchanges values stored in two locations.” --> It is note worthy that self-move-assignment is only required in the case where the object has already been moved-from. Self-swap does appear in some older standard library implementations of `std::random_shuffle()`. This underscores the need to support self-swap by the standard requirements.
+
+Supporting self-move-assignment for this narrow case imposes some additional complexity because `a = move(a)` is, in general, a contradiction <!-- I disagree, but regardless, even if we never come into agreement this point doesn't help your case and is a needless opportunity for contention.  I suggext omitting it. --> and is not required by the implementation of any standard component. The implementation required postcondition of `a = move(b)` is that `a` holds the prior value of `b` and the value of `b` is unspecified, but may be guaranteed to be a specific value. For example, if `a` and `b` are both of type `my_unique_ptr<T>` with the guarantee that `a` will hold the prior value of `b`, and `b` will be equal to `nullptr`. Then for the expression `a = move(a)`, the only way both of those guarantees could be satisfied is if `a` is already equal to `nullptr`. The current standard avoids this contradiction by defining the semantics of move assignment for `std::unique_ptr<T>` as equivalent to `reset(r.release())` which provides a stronger guarantee than any standard component implementation requires while satisfying the Standard requirements.
 
-Supporting self-move-assignment for this narrow case imposes some additional complexity because `a = move(a)` is, in general, a contradiction and is not required by the implementation of any standard component. The implementation required postcondition of `a = move(b)` is that `a` holds the prior value of `b` and the value of `b` is unspecified, but may be guaranteed to be a specific value. For example, if `a` and `b` are both of type `my_unique_ptr<T>` with the guarantee that `a` will hold the prior value of `b`, and `b` will be equal to `nullptr`. Then for the expression `a = move(a)`, the only way both of those guarantees could be satisfied is if `a` is already equal to `nullptr`. The current standard avoids this contradiction by defining the postcondition of move assignment for `std::unique_ptr<T>` as equivalent to `reset(r.release())` which provides a stronger guarantee than any standard component implementation requires while satisfying the Standard requirements.
+<!-- Doesn't this simply indicate that the guarantee you wrote for my_unique_ptr is wrongly phrased unless we add a special case for &a == &b? IMO you're not (yet) clearly pointing at an actual problem here. -->
```

Comment:
I'll look to rephrase this but I don't want to have to carve out an exception for self-move-assignment. It doesn't make sense to have a guarantee and then an exception for the aliased case. We have a guarantee for the lhs of assignment, and frequently we have a guaranteed value for a moved-from-object. Carving out a special case for self-move-assignment is just a hack to avoid the contradiction.

Reply:
🤷 “It's just a hack” seems to be in the eye of the beholder.  

The standard as currently written is intended to allow move self-assignment; that seems pretty clear from lots of things, including the `unique_ptr` semantics you note elsewhere.  Whether that was a good idea or not is a separate question that I don't want to grapple with; it's probably too late to take that decision back anyhow.  In a world where move self-assignment is allowed, if you are going to give postconditions for the rhs of move-assignment, you have to carve out a special case for self-assignment; that's just another rule of programming in C++, which is a consequence of the language design, like the rule of 5.

You do realize I'm not suggesting you propose carving out an exception in the standard, right?